File: mac/Sources/PixShell/Store/DownloadTasks.swift
import Foundation

/// 下载任务清单（工具浮窗「下载任务将显示在这里」那一块的数据源）。
/// 老仓库 #toolsDlBox 就是这个用途 —— 工具浮窗本体主要就是给用户看下载进度/结果的。
/// 只在内存里保留最近若干条；进程退出即清空（与老仓库一致，不做持久化）。
final class DownloadTasks {
    static let shared = DownloadTasks()
    private init() {}

    enum State { case running, done, failed }

    struct Task {
        let id: UUID
        var name: String        // 文件名/包名
        var dest: String        // 落地路径
        var state: State
        var detail: String      // 失败原因 / 完成提示
        var started: Date
    }

    private(set) var tasks: [Task] = []
    private let maxKeep = 30

    /// 有变化就通知（工具浮窗订阅它刷新列表）。
    var onChange: (() -> Void)?

    @discardableResult
    func start(name: String, dest: String) -> UUID {
        let t = Task(id: UUID(), name: name, dest: dest, state: .running, detail: "", started: Date())
        let id = t.id
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.tasks.insert(t, at: 0)
            if self.tasks.count > self.maxKeep { self.tasks.removeLast(self.tasks.count - self.maxKeep) }
            self.notify()
        }
        return id
    }

    func finish(_ id: UUID, ok: Bool, detail: String = "") {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            guard let i = self.tasks.firstIndex(where: { $0.id == id }) else { return }
            self.tasks[i].state = ok ? .done : .failed
            self.tasks[i].detail = detail
            self.notify()
        }
    }

    func clear() {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.tasks.removeAll()
            self.notify()
        }
    }

    private func notify() {
        if Thread.isMainThread { onChange?() }
        else { DispatchQueue.main.async { [weak self] in self?.onChange?() } }
    }
}

File: win/Store/DownloadTasks.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace PixShell;

/// <summary>
/// 下载任务清单（工具浮窗「下载任务将显示在这里」那一块的数据源）。
/// 老仓库 #toolsDlBox 就是这个用途 —— 工具浮窗本体主要就是给用户看下载进度/结果的。
/// 只在内存里保留最近若干条；进程退出即清空（与老仓库一致，不做持久化）。
/// 对齐 mac Store/DownloadTasks.swift。
/// </summary>
public static class DownloadTasks
{
    public enum State { Running, Done, Failed }

    public sealed class Task
    {
        public Guid Id { get; init; } = Guid.NewGuid();
        public string Name { get; set; } = "";
        public string Dest { get; set; } = "";
        public State State { get; set; } = State.Running;
        public string Detail { get; set; } = "";
        public DateTime Started { get; init; } = DateTime.Now;
    }

    private static readonly List<Task> _tasks = new();
    private const int MaxKeep = 30;

    /// <summary>有变化就通知（工具浮窗订阅它刷新列表）。</summary>
    public static event Action? Changed;

    public static IReadOnlyList<Task> Items
    {
        get { lock (_tasks) return _tasks.ToList(); }
    }

    public static Guid Start(string name, string dest)
    {
        var t = new Task { Name = name, Dest = dest, State = State.Running };
        lock (_tasks)
        {
            _tasks.Insert(0, t);
            if (_tasks.Count > MaxKeep) _tasks.RemoveRange(MaxKeep, _tasks.Count - MaxKeep);
        }
        Notify();
        return t.Id;
    }

    public static void Finish(Guid id, bool ok, string detail = "")
    {
        lock (_tasks)
        {
            var t = _tasks.FirstOrDefault(x => x.Id == id);
            if (t == null) return;
            t.State = ok ? State.Done : State.Failed;
            t.Detail = detail;
        }
        Notify();
    }

    public static void Clear()
    {
        lock (_tasks) _tasks.Clear();
        Notify();
    }

    private static void Notify()
    {
        var app = System.Windows.Application.Current;
        if (app == null) { Changed?.Invoke(); return; }
        // UI 线程外（SFTP 传输在后台线程回调）也要能安全刷新界面。
        if (app.Dispatcher.CheckAccess()) Changed?.Invoke();
        else app.Dispatcher.BeginInvoke(new Action(() => Changed?.Invoke()));
    }
}
